packages/common/cumulus-api-client/CumulusApiClient.js

```diff
@@ -0,0 +1,155 @@
+  async _getAuthTokenRecord() {
+    const params = {
+      TableName: this.config.authTokenTable,
+      Key: {
+        tokenAlias: this.config.tokenSecretName
+      }
+    };
+    const tokenResponse = await dynamodbDocClient().get(params).promise();
+    if (!tokenResponse.Item) {
+      throw new CumulusAuthTokenError(`No bearer token with alias '${this.config.tokenSecretName}'
+      found in ${this.config.authTokenTable}`);
+    }
+    return decryptBase64String(tokenResponse.Item.bearerToken);
+  }
+
+  /**
+   *
+   * @param {string} token - Updates the row at config.tokenSecretName with
+   *                         a kms encrypted token record
+   * @returns {Promise<Object>} - dynamodbDocClient response
+   */
+  async _updateAuthTokenRecord(token) {
+    const encryptedToken = await encrypt(this.config.kmsId, token);
+    const params = {
+      TableName: this.config.authTokenTable,
+      Key: {
+        tokenAlias: this.config.tokenSecretName
+      },
+      UpdateExpression: 'set bearerToken = :t',
+      ExpressionAttributeValues: {
+        ':t': encryptedToken
+      }
+    };
+    return dynamodbDocClient().update(params).promise();
+  }
+
+  /**
+   * Helper function to check how much time the JWT token has left before expiration
+   * @param {*} token - bearer token
+   *
+   * @returns {number} - the number of seconds left before the token expires
+   */
+  async _getTokenTimeLeft(token) {
+    return (((decode(token).exp) - (Date.now() / 1000)));
+  }
+
+  /**
+   * Validates if a token is close to expiration, and throws an error if so
+   * @param {*} token
+   *
+   * @throws (CumulusAuthTokenError)
+   */
+  async _validateTokenExpiry(token) {
+    const tokenSecondsRemaining = await this._getTokenTimeLeft(token);
+    if (tokenSecondsRemaining <= 0) {
+      throw new CumulusAuthTokenError('Token expired, obtraining new token');
+    }
+    if (tokenSecondsRemaining <= this.config.tokenExpireBuffer && tokenSecondsRemaining > 0) {
+      throw new CumulusAuthTokenError('Token nearing experation, obtraining new token');
+    }
+  }
+
+  /**
+   * Gets a token using the CumulusApiClient caching scheme:
+   *  1) Attempts to get a valid token from the cache table
+   *  2) If it is successful, check (if possible) if the token is about to expire
+   *  3) If the token does not exist, or is about to expire, generate a new token
+   *     and store it inthe cache
+   *
+   * @returns {string} - returns active valid bearer token
+   */
+  async getCacheAuthToken() {
+    let token;
+    try {
+      token = await this._getAuthTokenRecord();
+      await this._validateTokenExpiry(token);
+      return token;
+    } catch (error) {
+      if (error.name === 'CumulusAuthTokenError') {
+        logger.info('API Client access token expired, generating new token');
+        // We're not refreshing as refreshing invalidates what could be
+        // an active key.
+        const updateToken = await this.createNewAuthToken();
```

Comment:
We're not hitting /token/refresh as a successful refresh gets a new token, but causes the original key to be removed from the db and/or expired.     As lambdas are running asynchronously this would cause potential badness, depending on implementation.